topic_tools/src/demux_node.cpp

```diff
@@ -0,0 +1,186 @@
+// Copyright 2024 Rufus Wong
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "topic_tools/demux_node.hpp"
+
+#include <memory>
+#include <optional>  // NOLINT : https://github.com/ament/ament_lint/pull/324
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "rclcpp/rclcpp.hpp"
+
+namespace topic_tools
+{
+DemuxNode::DemuxNode(const rclcpp::NodeOptions & options)
+: ToolBaseNode("demux", options)
+{
+  using std::placeholders::_1;
+  using std::placeholders::_2;
+
+  output_topic_ = declare_parameter("initial_topic", "");
+  input_topic_ = declare_parameter("input_topic", "~/input");
+  lazy_ = false;
+  output_topics_ = declare_parameter<std::vector<std::string>>("output_topics");
+  if (output_topic_.empty()) {
+    output_topic_ = output_topics_.front();
+  }
```

Comment:
Assigning `output_topic_ = output_topics_.front();` when `output_topic_.empty()` seems not a good idea, since `front()` will be pointing out to the ambient address in memory.
According to the https://en.cppreference.com/w/cpp/container/vector/front
> Calling front on an empty container causes undefined behavior.

Reply:
it's a bit confusing, but it's assigning something to `output_topic_` if it's empty (`output_topic_.empty()`). It's not using `output_topics_.front()` when `output_topics_.empty()`.

Reply:
Yeah, To avoid confusion it would be better to have 3 variables `initial_topic`, `output_topics` and final `output_topic_`.
Also still need to check that `(!output_topics_.empty())` before using `output_topic_.front()`

Reply:
Added `initial_topic` variable for both mux and demux